connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java

```diff
@@ -63,24 +71,58 @@
     private static final Properties DEFAULT_BROKER_CONFIG = new Properties();
     private static final String REST_HOST_NAME = "localhost";
 
-    private final Connect[] connectCluster;
+    private final List<Connect> connectCluster;
     private final EmbeddedKafkaCluster kafkaCluster;
     private final Map<String, String> workerProps;
     private final String connectClusterName;
     private final int numBrokers;
+    private final int numInitialWorkers;
+    private final boolean failOnUngracefulShutdown;
 
-    private EmbeddedConnectCluster(String name, Map<String, String> workerProps, int numWorkers, int numBrokers, Properties brokerProps) {
+    private EmbeddedConnectCluster(String name, Map<String, String> workerProps, int numWorkers,
+                                   int numBrokers, Properties brokerProps,
+                                   boolean failOnUngracefulShutdown) {
         this.workerProps = workerProps;
         this.connectClusterName = name;
         this.numBrokers = numBrokers;
         this.kafkaCluster = new EmbeddedKafkaCluster(numBrokers, brokerProps);
-        this.connectCluster = new Connect[numWorkers];
+        this.connectCluster = new ArrayList<>();
+        this.numInitialWorkers = numWorkers;
+        this.failOnUngracefulShutdown = failOnUngracefulShutdown;
     }
 
+    /**
+     * A more graceful way to handle abnormal exit of services in integration tests.
+     */
+    public Exit.Procedure exitProcedure = (code, message) -> {
+        if (code != 0) {
+            String exitMessage = "Abrupt service exit with: " + code + " and " + message;
+            log.warn(exitMessage);
```

Comment:
that's the logger of the test right?

Reply:
Yeah, it doesn't seem like this would be a problem as long as `EmbeddedConnectCluster` is used in tests, which is currently the intended scope.

Reply:
yeah, I was bringing up this up, since there are some System.exit(1) calls coming from the DistributedHerder. I don't think it matters who owns the logger at that point. The entire jvm should shutdown. having said that, I think the shutdown hooks were problematic in log4j2. so we should be fine. also, you are overriding the exit procedure here, so it should be fine.